### ISSUE TITLE
Cluster Templates - RKE2 Example Template

### ISSUE THREAD

**Issue by maggieliu** [rancher/rancher#30099]
Make the cluster template generic for all cluster type.

**Comment by cloudnautique** [comment]
An example template for AWS and vSphere should include:

- Multiple node pools for etcd/control plane and workers.
- Prespecified Cloud credential (Admin will provide)
- Allow users to enable A.C.E and specify configuration options.
- Install Rancher v2 monitoring
- Assign cluster members (base operators)

A consumer of the template should be able to specify:
- Kubernetes version
- FQDN of ACE and custom certificate (optional)
- Labels for the cluster object (to auto-join Rancher CD cluster groups)
- Additional cluster members
- AWS vpc/subnets from drop down.
- vSphere datacenter from a drop down

**Comment by deniseschannon** [comment]
Chart Annotation to indicate this is a cluster template

```
catalog.cattle.io/type=cluster
```

**Comment by cbron** [comment]
This will also need the hidden annotation because we don't want people attempting to use it via apps&marketplace

**Comment by deniseschannon** [comment]
Spoke with @vincent99 and we should *not* use the hidden annotation cause that causes charts to be hidden everywhere. 

FYI, 

Apps & Marketplace will hide all charts that have a `catalog.cattle.io/type=*` annotation, so by adding the cluster-template annotation, it will automatically be hidden.

**Comment by StrongMonkey** [comment]
The first POC version is implemented in [here](https://github.com/StrongMonkey/cluster-template-examples#rke2-cluster-template). Have two general questions:
1. Should the chart live in rancher/charts instead of rancher/cluster-template-example?
2. How UI is going to consume it?(populate node pools values, version control and rollback)

**Comment by StrongMonkey** [comment]
Feature checklist according to Bill:

- [ ] Multiple node pools for etcd/control plane and workers.
- [ ] Prespecified Cloud credential (Admin will provide)
- [ ] Allow users to enable A.C.E and specify configuration options.
- [ ] Install Rancher v2 monitoring (Require api changes)
- [ ] Assign cluster members (base operators) (Require api changes)
- [ ] Kubernetes version
- [ ] FQDN of ACE and custom certificate (optional)
- [ ] Labels for the cluster object (to auto-join Rancher CD cluster groups)
- [ ] Additional cluster members (Require api changes)
- [ ] AWS vpc/subnets from drop down. (Require UI integration)
- [ ] vSphere datacenter from a drop down (Require UI integration)

**Comment by deniseschannon** [comment]
This is ready to test with the example.

Note: The cloud credential can only be selected with the correct secrete name. 

There is an alternative issue to make the UX friendly

**Comment by slickwarren** [comment]
https://github.com/rancher/dashboard/issues/2972

**Comment by slickwarren** [comment]
I did a smoke test on Friday using master-head + the example from above. Everything seemed to work using digital ocean + default settings. I will start going through the detailed tests this week

**Comment by slickwarren** [comment]
UI Ticket: https://github.com/rancher/dashboard/issues/2972
tested on master-head (4d1840056b02e390c29a9497abffc840c1de3b28):
* add/remove/edit RKE2 templates -- pass
* Provision/edit/remove cluster using RKE2 templates -- fail (see UI ticket for details)
  * via UI in rancher -- pass
  * via yaml in rancher -- pass
  * Digital ocean -- pass
  * hosted rancher -- pass
  * all others fail (vsphere, azure, custom, ec2)
  * use invalid settings, then edit with valid settings -- pass 
  * v2charts unable to install https://github.com/rancher/rancher/issues/33339
Blocked on the following:
* backup & restore -- fail https://github.com/rancher/rancher/issues/32599
* RBAC -- standard user -- blocked by https://github.com/rancher/rancher/issues/32414
* hardened cluster testing -- rke2 hardened clusters provisioned through rancher isn't a supported use case yet (tbd)
* airgap testing -- lower priority, takes more time from QA to setup resources, will test this when other blockers are looked in to
* fleet testing -- still following up with engineering to understand this use case for rke2 templates
* windows -- rke2 is not working on windows yet

**Comment by slickwarren** [comment]
on 2.6.0-rc1: 
UI Ticket: https://github.com/rancher/dashboard/issues/2972
tested on master-head (4d1840056b02e390c29a9497abffc840c1de3b28):
* add/remove/edit RKE2 templates -- pass
* Provision/edit/remove cluster using RKE2 templates -- pass
  * via UI in rancher -- pass
  * via yaml in rancher -- pass
  * Digital ocean -- pass
  * hosted rancher -- pass
  * ec2, custom -- pass
  * use invalid settings, then edit with valid settings -- pass 
  * v2charts - monitoring v2 -- pass
  * deploy template through fleet -- pass

Blocked on the following:
* backup & restore -- blocked https://github.com/rancher/rancher/issues/32599
* RBAC -- standard user -- blocked by https://github.com/rancher/rancher/issues/32414
* hardened cluster testing -- rke2 hardened clusters provisioned through rancher isn't a supported use case yet (tbd)
* airgap testing -- lower priority to test, but should work

untested: 
* windows -- rke2 is not working [on windows yet](https://github.com/rancher/rancher/issues/33625)
* vsphere/azure providers
* use of private registry/proxy vars via template

**Comment by vincent99** [comment]
Those are not happening in 2.6.3... catalog questions independent of each other so even if written it is not clear how a question for a VPC or similar would know what credential to use to load the list.

**Comment by slickwarren** [comment]
opened a PR to support harvester https://github.com/rancher/cluster-template-examples/pull/4

**Comment by slickwarren** [comment]
we have some open tickets, but the basic use cases are working for rke2 templates. Here is a writeup of status as of 2.6.4-rc4:
* chart deployments are not yet supported https://github.com/rancher/rancher/issues/33339
* basic cluster deployment for all default-enabled node drivers, and custom clusters is supported
  * digital ocean
  * aws
  * azure
  * linode -- supported, but not implemented in the example chart
  * harvester (pr above for implementation)
  * vsphere
* questions.yaml does not yet support multiple node pools https://github.com/rancher/dashboard/issues/5236
  * configuring multiple node pools is supported in the UI via the `edit as yaml` tab

**Comment by slickwarren** [comment]
closing in favor of tickets mentioned above
